base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1604 @@
+
+`iridx` is the inverse function of `irval` (not formally), i.e.
+`iridx(irval(xidx, state), state) === xidx`.
+"""
+function irval(xidx::Int, estate::EscapeState)
+    x = xidx > estate.nargs ? SSAValue(xidx-estate.nargs) : Argument(xidx)
+    return x
+end
+
+function getaliases(x::Union{Argument,SSAValue}, estate::EscapeState)
+    xidx = iridx(x, estate)
+    aliases = getaliases(xidx, estate)
+    aliases === nothing && return nothing
+    return Union{Argument,SSAValue}[irval(aidx, estate) for aidx in aliases]
+end
+function getaliases(xidx::Int, estate::EscapeState)
+    aliasset = estate.aliasset
+    root = find_root!(aliasset, xidx)
+    if xidx ≠ root || aliasset.ranks[xidx] > 0
+        # the size of this alias set containing `key` is larger than 1,
+        # collect the entire alias set
+        aliases = Int[]
+        for aidx in 1:length(aliasset.parents)
+            if aliasset.parents[aidx] == root
+                push!(aliases, aidx)
+            end
+        end
+        return aliases
+    else
+        return nothing
+    end
+end
+
+isaliased(x::Union{Argument,SSAValue}, y::Union{Argument,SSAValue}, estate::EscapeState) =
+    isaliased(iridx(x, estate), iridx(y, estate), estate)
+isaliased(xidx::Int, yidx::Int, estate::EscapeState) =
+    in_same_set(estate.aliasset, xidx, yidx)
+
+"""
+    EscapeLatticeCache(x::EscapeLattice) -> x′::EscapeLatticeCache
+
+The data structure for caching `x::EscapeLattice` for interprocedural propagation,
+which is slightly more efficient than the original `x` object.
+"""
+struct EscapeLatticeCache
+    AllEscape::Bool
+    ReturnEscape::Bool
+    ThrownEscape::Bool
+    function EscapeLatticeCache(x::EscapeLattice)
+        x === ⊤ && return new(true, true, true)
+        ThrownEscape = isempty(x.ThrownEscape) ? false : true
+        return new(false, x.ReturnEscape, ThrownEscape)
+    end
+end
+
+"""
+    cache_escapes!(linfo::MethodInstance, estate::EscapeState, _::IRCode)
+
+Transforms escape information of `estate` for interprocedural propagation,
+and caches it in a global cache that can then be looked up later when
+`linfo` callsite is seen again.
+"""
+function cache_escapes! end
+
+# when working outside of Core.Compiler, cache as much as information for later inspection and debugging
+if _TOP_MOD !== Core.Compiler
+    struct EscapeCache
+        cache::Vector{EscapeLatticeCache}
+        state::EscapeState # preserved just for debugging purpose
+        ir::IRCode         # preserved just for debugging purpose
+    end
+    const GLOBAL_ESCAPE_CACHE = IdDict{MethodInstance,EscapeCache}()
+    function cache_escapes!(linfo::MethodInstance, estate::EscapeState, cacheir::IRCode)
+        cache = EscapeCache(to_interprocedural(estate), estate, cacheir)
+        GLOBAL_ESCAPE_CACHE[linfo] = cache
+        return cache
+    end
+    argescapes_from_cache(cache::EscapeCache) = cache.cache
+else
+    const GLOBAL_ESCAPE_CACHE = IdDict{MethodInstance,Vector{EscapeLatticeCache}}()
```

Comment:
The escape information should move inside CodeInstance, then I assume we wouldn't need a separate cache?

Reply:
DONE.